src/tail/log_server.rs

```diff
@@ -31,54 +31,43 @@ impl LogServer {
     }
 
     pub async fn run(self) -> Result<()> {
-        // this is so bad
-        // but i also am so bad at types
-        // TODO: make this less terrible
-        match self.format.as_str() {
-            "json" => {
-                let service = make_service_fn(|_| async {
-                    Ok::<_, hyper::Error>(service_fn(print_logs_json))
-                });
-
-                let server = self.server.serve(service);
-
-                // The shutdown receiver listens for a one shot message from our sigint handler as a signal
-                // to gracefully shut down the hyper server.
-                let shutdown_rx = self.shutdown_rx;
-
-                let graceful = server.with_graceful_shutdown(async {
-                    shutdown_rx.await.ok();
-                });
-
-                graceful.await?;
-
-                Ok(())
+        let format = self.format.clone(); // Having to clone this so much is a little gross TODO
```

Comment:
I tried that but the lifetime for the closure for `service_fn` is `'static` meaning it'll outlive `self.format` if I try to take it as a reference. I might not fully understand what is going on though

Reply:
Yeah I had this same problem yesterday - you can avoid it by using manual impls instead of a service function:

```rust
impl<T> Service<T> for MakeFooService {
    // ....

    fn poll_ready(&mut self, _: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
        Poll::Ready(Ok(()))
    }

    fn call(&mut self, _: T) -> Self::Future {
        let service = self.0.clone();
        ok(service)
    }
}

pub struct FooService(Foo);

impl Service<Request<Body>> for FooService {
    // ...

    fn poll_ready(&mut self, _: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
        Poll::Ready(Ok(()))
    }

    fn call(&mut self, req: Request<Body>) -> Self::Future {
        handler(req, self.0.clone())
    }
}
```
Or something like that anyway.

Reply:
@jspspike this would be cool but also it's not a huge deal -- if you wanna do this go ahead, otherwise it's not that bad. Just add a comment explaining why we need to keep cloning.

Reply:
I un-nested the closures and made it a bit cleaner, I think it's in a good state now